imap/imap.go

```diff
@@ -0,0 +1,365 @@
+	client, err := mbox.newClient()
+	if err != nil {
+		return err
+	}
+	defer func() {
+		client.Close(true)
+		client.Logout(30 * time.Second)
+	}()
+	for _, u := range uids {
+		err := alterEmail(client, u, "\\SEEN", false)
+		if err != nil {
+			return err //return on first failure
+		}
+	}
+	return nil
+
+}
+
+// DeleteEmails will delete emails from the supplied slice of UIDs
+//func DeleteEmails(info Mailbox, uids []uint32) error {
+func (mbox *Mailbox) DeleteEmails(uids []uint32) error {
+	client, err := mbox.newClient()
+	if err != nil {
+		return err
+	}
+	defer func() {
+		client.Close(true)
+		client.Logout(30 * time.Second)
+	}()
+	for _, u := range uids {
+		err := deleteEmail(client, u)
+		if err != nil {
+			return err //return on first failure
+		}
+	}
+	return nil
+
+}
+
+// Validate validates supplied IMAP model by connecting to the server
+func Validate(s *models.IMAP) error {
+
+	err := s.Validate()
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+
+	s.Host = s.Host + ":" + strconv.Itoa(int(s.Port)) // Append port
+	mailServer := Mailbox{
+		Host:   s.Host,
+		TLS:    s.TLS,
+		User:   s.Username,
+		Pwd:    s.Password,
+		Folder: s.Folder}
+
+	client, err := mailServer.newClient()
+	if err != nil {
+		log.Error(err.Error())
+	} else {
+		client.Close(true)
+		client.Logout(30 * time.Second)
+	}
+	return err
+}
+
+// Response is a helper struct to wrap the email responses and possible errors.
+type Response struct {
+	Email Email
+	Err   error
+}
+
+// newClient will initiate a new IMAP connection with the given creds.
+//func newClient(info Mailbox) (*imap.Client, error) {
+func (mbox *Mailbox) newClient() (*imap.Client, error) {
+	var client *imap.Client
+	var err error
+	if mbox.TLS {
+		client, err = imap.DialTLS(mbox.Host, new(tls.Config))
+		if err != nil {
+			return client, err
+		}
+	} else {
+		client, err = imap.Dial(mbox.Host)
+		if err != nil {
+			return client, err
+		}
+	}
+
+	_, err = client.Login(mbox.User, mbox.Pwd)
+	if err != nil {
+		return client, err
+	}
+
+	_, err = imap.Wait(client.Select(mbox.Folder, mbox.ReadOnly))
+	if err != nil {
+		return client, err
+	}
+
+	return client, nil
+}
+
+const dateFormat = "02-Jan-2006"
+
+// findEmails will run a find the UIDs of any emails that match the search.:
+func findEmails(client Client, search string, since *time.Time) (*imap.Command, error) {
+	var specs []imap.Field
+	if len(search) > 0 {
+		specs = append(specs, search)
+	}
+
+	if since != nil {
+		sinceStr := since.Format(dateFormat)
+		specs = append(specs, "SINCE", sinceStr)
+	}
+
+	// get headers and UID for UnSeen message in src inbox...
+	cmd, err := imap.Wait(client.UIDSearch(specs...))
+	if err != nil {
+		return &imap.Command{}, fmt.Errorf("uid search failed: %s", err)
+	}
+	return cmd, nil
+}
+
+var GenerateBufferSize = 100
+
+//func generateMail(info Mailbox, search string, since *time.Time, markAsRead, delete bool) (chan Response, error) {
```

Comment:
Small nit - looks like some commented code slipped in here.

Reply:
Done